Kernel/Net/IPv4Socket.cpp

```diff
@@ -465,50 +465,100 @@ KResult IPv4Socket::getsockopt(FileDescription& description, int level, int opti
 int IPv4Socket::ioctl(FileDescription&, unsigned request, unsigned arg)
 {
     REQUIRE_PROMISE(inet);
-    auto* ifr = (ifreq*)arg;
-    if (!Process::current->validate_read_typed(ifr))
-        return -EFAULT;
 
-    char namebuf[IFNAMSIZ + 1];
-    memcpy(namebuf, ifr->ifr_name, IFNAMSIZ);
-    namebuf[sizeof(namebuf) - 1] = '\0';
-    auto adapter = NetworkAdapter::lookup_by_name(namebuf);
-    if (!adapter)
-        return -ENODEV;
+    auto ioctl_route = [request, arg]() {
+        auto* route = (rtentry*)arg;
+        if (!Process::current->validate_read_typed(route))
+            return -EFAULT;
 
-    switch (request) {
-    case SIOCSIFADDR:
-        if (!Process::current->is_superuser())
-            return -EPERM;
-        if (ifr->ifr_addr.sa_family != AF_INET)
-            return -EAFNOSUPPORT;
-        adapter->set_ipv4_address(IPv4Address(((sockaddr_in&)ifr->ifr_addr).sin_addr.s_addr));
-        return 0;
+        char namebuf[IFNAMSIZ + 1];
+        memcpy(namebuf, route->rt_dev, IFNAMSIZ);
+        namebuf[sizeof(namebuf) - 1] = '\0';
+
+        auto adapter = NetworkAdapter::lookup_by_name(namebuf);
+        if (!adapter)
+            return -ENODEV;
+
+        switch (request) {
+        case SIOCADDRT:
+            if (!Process::current->is_superuser())
+                return -EPERM;
+            if (route->rt_gateway.sa_family != AF_INET)
+                return -EAFNOSUPPORT;
+            if ((route->rt_flags & (RTF_UP | RTF_GATEWAY)) != (RTF_UP | RTF_GATEWAY))
+                return -EINVAL; // FIXME: Find the correct value to return
+            adapter->set_ipv4_gateway(IPv4Address(((sockaddr_in&)route->rt_gateway).sin_addr.s_addr));
```

Comment:
This will remove any previously existing route. We really need a separate (global) routing table in the kernel that's separate from the `NetworkAdapter` class.

Something like this might work for starters:

```cpp
struct Route {
    IPv4Address destination;
    IPv4Address gateway;
    IPv4Address netmask;
    WeakPtr<NetworkAdapter> adapter;
};
Vector<Route> m_routes;
```

I'll merge this patch since it's still improving the current situation. But then we shouldn't build more stuff on top of the "only one route per NetworkAdapter" concept :)